Gems/Atom/Utils/Code/Include/Atom/Utils/StableDynamicArray.inl

```diff
@@ -225,18 +204,20 @@ namespace AZ
     StableDynamicArrayMetrics StableDynamicArray<T, ElementsPerPage, Allocator>::GetMetrics()
     {
         StableDynamicArrayMetrics metrics;
-        Page* page = m_firstPage;
 
-        while (page)
+        for (Page* page : m_pages)
```

Comment:
The best ideas I have come up with so far for m_pages resizing while accessing via weak handle are:
1. Use a 65536 element fixed_vector (fixed size of 512kb for 64-bit pointers) so there's no resizing.
2. Keep an extra temporary page vectors around, with an atomic bool that determines which one to use for random access via a weak handle.
During insert, do
```
if (m_pages.size() == m_pages.capacity())
{
    m_tempPages = m_pages;
    m_useTempPages = true;
}
m_pages.push_back(page);
m_useTempPages = false;
```

During GetData, do
```
return m_useTempPages ? *m_tempPages[handle.m_pageIndex]->GetItem(handle.m_elementIndex)
                      : *m_pages[handle.m_pageIndex]->GetItem(handle.m_elementIndex);
```

Then in the case where there is a resize, the temp vector is prepared, all random access starts going through the temp vector, and once the resize is complete, it starts going back through the regular pages vector.

Reply:
oof, that copy of 512kb though.

Reply:
I like the idea of using a fixed_vector, it complete avoids the issue of reallocating the vector. That plus the concurrency checker should let us know if we have an issue we need to fix.

The case that still seems like it could be an issue is 2 threads trying to add at the same time, but we may just be able to use a lock for that.

Reply:
2 Thread adding at the same time is not something I'm trying to support, and that's protected higher up with a lock.

Current usage doesn't require it to have thread safe access while inserting, but I was leveraging that for instancing. In theory, that's the benefit you get from it being stable. During MeshFeatureProcessor::Simulate, one thread could be running Init, which is inserting into the MeshInstanceManager, while another thread could be accessing from the MeshInstanceManager (which I backed with my own PagedDataVector in my prototype, but I'd like to switch to StableDynamicArray). So that's where I'm inserting and accessing at the same time.

Reply:
I can potentially relax the need for threaded access while inserting. I already ended up splitting the MeshFeatureProcessor::Simulate into stages, where the init jobs needed to finish before a per-instance-group job, which needed to finish before an update culling data job. So I could potentially stage that work further. But Init itself is both inserting and accessing data, so I'd need to split that up which would be an even more intrusive change I'd like to avoid for now.